File: packages/cli/src/lib/logging.ts
export type LogRoute =
  | { readonly type: "root" }
  | { readonly type: "subdomain"; readonly subdomain: string };

export type DerivePublicUrlOptions = {
  readonly serverUrl: string;
  readonly subdomain?: string;
};

export type RuntimeLogger = {
  info(message: string): void;
  error(message: string): void;
};

export const sanitizeLogPath = (url?: string): string => {
  if (!url) {
    return "/";
  }

  try {
    return new URL(url, "http://proxer.local").pathname || "/";
  } catch {
    return "/";
  }
};

export const sanitizeLogUrl = (url: string): string => {
  try {
    const parsed = new URL(url);
    const hostIndex = url.indexOf(parsed.host);
    const afterHost =
      hostIndex >= 0 ? url.slice(hostIndex + parsed.host.length) : "";
    const hasExplicitPath = afterHost.startsWith("/");

    return `${parsed.protocol}//${parsed.host}${hasExplicitPath ? parsed.pathname : ""}`;
  } catch {
    return url.split("?")[0]?.split("#")[0] ?? "";
  }
};

export const derivePublicUrl = ({
  serverUrl,
  subdomain,
}: DerivePublicUrlOptions): string => {
  const url = new URL(serverUrl);

  if (url.protocol === "wss:") {
    url.protocol = "https:";
  } else if (url.protocol === "ws:") {
    url.protocol = "http:";
  }

  url.username = "";
  url.password = "";
  url.pathname = "/";
  url.search = "";
  url.hash = "";

  if (subdomain) {
    url.hostname = `${subdomain}.${url.hostname}`;
  }

  return url.origin;
};

export const formatRoutePrefix = (route?: LogRoute): string => {
  if (!route) {
    return "[unknown]";
  }

  return route.type === "root" ? "[root]" : `[${route.subdomain}]`;
};

File: packages/cli/lib/src/util/logging.dart
class LogRoute {
  const LogRoute.root() : subdomain = null;
  const LogRoute.subdomain(this.subdomain);

  final String? subdomain;
}

String sanitizeLogPath(String value) {
  final uri = Uri.tryParse(value);
  return uri?.path.isNotEmpty == true ? uri!.path : '/';
}

String sanitizeLogUrl(String value) {
  final uri = Uri.tryParse(value);
  if (uri == null) {
    return value;
  }
  return uri.replace(userInfo: '', query: null, fragment: null).toString();
}

String derivePublicUrl({required String serverUrl, String? subdomain}) {
  final uri = Uri.parse(serverUrl);
  final scheme = uri.scheme == 'wss' || uri.scheme == 'https'
      ? 'https'
      : 'http';
  final host = subdomain == null ? uri.host : '$subdomain.${uri.host}';
  return uri
      .replace(scheme: scheme, host: host, path: '', query: null)
      .toString();
}

String formatRoutePrefix(LogRoute? route) {
  if (route == null) {
    return '';
  }
  return route.subdomain == null ? '[root]' : '[${route.subdomain}]';
}
